content/courses/Ada_For_The_C_Embedded_Developer/chapters/05_SPARK.rst

```diff
@@ -46,110 +115,139 @@ frames. For example:
 
     procedure Some_Process is
-       A : Arr (1 .. 100);
+       L : constant List (1 .. 100) := (others => 42);
     begin
-       Put_Line (Integer'Image (Read (A, 1, 10)));
+       Put_Line (Integer'Image (Value (L, 1, 10)));
     exception
        when Constraint_Error =>
-          Put_Line ("FAILURE");
-    end;
-
-.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.SPARK.Exceptions
+          Put_Line ("Constraint_Error caught in Some_Process");
+          Put_Line ("Some_Process completes normally");
+    end Some_Process;
 
     with Some_Process;
+    with Ada.Text_IO; use Ada.Text_IO;
 
     procedure Main is
     begin
        Some_Process;
+       Put_Line ("Main completes normally");
     end Main;
 
 :code-config:`accumulate_code=False`
 
 :code-config:`reset_accumulator=True`
 
-On the above code, :ada:`Some_Process` calls :ada:`Read` with values leading to
-an out-of-bound access (:ada:`1 + 10 * 10 = 101`, beyond :ada:`A` last index).
-This will trigger an exception in :ada:`Read` which will happen prior to
-reading the value in memory. :ada:`Read` doesn't have any code handling
-exceptions, so the exception is propagated to the caller :ada:`Some_Process`.
-:ada:`Some_Process` has an exception handled, introduced by the construction
-exception, which looks in particular for the exception of kind
-:ada:`Constraint_Error`. It happens so that :ada:`Constraint_Error` is the
-exception raised in case of array out of bound access, so the code of the
-handler will be executed (printing an error message on the screen) and the
-function :ada:`Some_Process` will then return normally.
-
-Full exception propagation as described above are not the norm for embedded
-applications, in particular when high level of integrity are required. The code
-needed for exception propagation can be rather complex and difficult to
-certify, and having control flow jumping through the call stack creates
-situations with inconsistent data (for example, if it happens in the middle of
-an update). Not to mention of course the fact that checking for the failing
-conditions is taking processor time in the first place, which hinders
-application performances. Instead of the full exception model, embedded
-application usually resolve to alternative options. The following options are
-mostly GNAT-specific, as the Ada standard doesn't provide much guidance there.
-
-The first one consists in deactivating exceptions altogether |mdash| or more
-precisely, deactivating checks, which will mean that the compiler will not
-generate verifications leading to exceptions in cases such as buffer overflow.
-Of course, this makes then the code vulnerable to attacks, unless otherwise
-verified (e.g. through static analysis). This can be done at the file level,
-through the ``-gnatp`` compiler switch, or locally to an area within the
-:ada:`pragma Suppress`. For example, we can write:
+Procedure :ada:`Main` calls :ada:`Some_Process`, which in turn calls 
```

Comment:
Editorial suggestion: this paragraph is actually a perfect description of how the exception is propagated and handled in the code above. And yet, I had to read it carefully to understand it. Maybe we could split the paragraph in bullet points for each subprogram in the chain? Another (more complicated) approach would be to add a graph that illustrates the chain and shows where the exception is raised, propagated and handled.